### DISCUSSION TITLE
[c_cpp][intellisense] include issues

### DISCUSSION THREAD

**Discussion by stark-dev** [microsoft/vscode-cpptools#9792] [Q&A] [answered]
Hi, first of all thanks for your work, I'm using `vscode` daily for all my development and it's becoming greater at every release.

I'm currently having some issue with the C/C++ intellisense. I have quite a big project with lot of dependencies, that are cloned locally in the build folder when configuring the cmake target.
I had no issues so far, until today the intellisense started to report more than 100 errors, including `cannot open source file`, `expected a ;`, etc.

The project is correctly built using the cmake extension and the only difference wrt yesterday is that I rebooted the machine used for development (Ubuntu 20.04).
I checked similar issues, which usually suggest to update the include path in the `c_cpp_configuration.json` file. I used simply the include path `${workspaceFolder}/**"`, which worked fine so far.
I tried to rebuild the intellisense database, kill the `cpptools` and `cpptools-srv` processes, delete the `ipch` folder in `.cache/vscode-cpptools`, disabled and re-enabled the extension. I also deleted and cloned the repository to a different location, but none of this steps helped with the issue.

I'm wondering if given the size of the project, the intellisense may have some issues to parse correctly all the files. Any suggestion in this regard?

Here is my current setup:
```
Version: 1.70.2
Commit: e4503b30fc78200f846c62cf8091b76ff5547662
Date: 2022-08-16T05:36:25.715Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Linux x64 5.15.0-46-generic
```

Thanks in advance for your help, let me know if I can provide any other information that could be useful to investigate this issue.

Best regards,

Mauro

**Comment by bobbrow** [comment]
While you have the file open with the errors in it, can you run the `C/C++: Log Diagnostics` command and share the output?  You can also run `CMake: Log Diagnostics` and share that too.  We can see if anything stands out in those two logs and then go from there.

**Reply by stark-dev** [reply]
Thanks a lot for your reply, in the mean time I run some other tests:
I cloned the repository from scratch, run the CMake configuration and build again. This time intellisense and auto completion are fine.
I noticed some interesting differences in the log files, the path of the compiler and absence of the include folders in the faulty one.

I had the same behavior once already, the workspace looks good for few weeks and then it starts to show all these errors.
I also attached my `c_cpp_properties.json` and the `settings.json` local configs, which are the same for both the good and the faulty workspaces.

Thanks again for your help.

BR,

Mauro

[c_cpp_diagnostics-faulty.log](https://github.com/microsoft/vscode-cpptools/files/9435137/c_cpp_diagnostics-faulty.log)
[c_cpp_diagnostics-good.log](https://github.com/microsoft/vscode-cpptools/files/9435138/c_cpp_diagnostics-good.log)
[cmake_diagnostics-faulty.log](https://github.com/microsoft/vscode-cpptools/files/9435139/cmake_diagnostics-faulty.log)
[cmake_diagnostics-good.log](https://github.com/microsoft/vscode-cpptools/files/9435140/cmake_diagnostics-good.log)
[c_cpp_properties.txt](https://github.com/microsoft/vscode-cpptools/files/9435144/c_cpp_properties.txt)
[settings.txt](https://github.com/microsoft/vscode-cpptools/files/9435145/settings.txt)

**Reply by bobbrow** [reply]
So it looks like in one case the CMake Tools extension sends the configuration to the C++ extension for IntelliSense, and in the other case it doesn't.

I see this in the faulty config:
```
    "requests": [
      "file:///home/user/Developer/Projects/MyProject/src/file.cpp"
    ],
    "responses": [],
    "partialMatches": [
      {
        "request": "file:///mnt/data/ws/MyProject/src/file.cpp",
        "matches": "/home/user/Developer/Projects/MyProject/src/file.cpp"
      }
    ],
```
I don't understand how the two `request` URI's could be different in the log because when we log these out, I don't see any processing happening, but what it does seem to indicate is that you are working with a symlink that we're getting an incorrect mapping for.

Can you confirm which way you open the folder in the faulty case?  With the symlink or with the real path?

**Reply by stark-dev** [reply]
Yes, in the faulty case the `Projects` folder is a symlink. It's odd that the issue appears after some time though, I always use the symlink path, I just made a test without symlink when cloning the repo again

**Reply by sean-mcmanus** [reply]
Sounds like https://github.com/microsoft/vscode-cpptools/issues/9498 and https://github.com/microsoft/vscode-cpptools/issues/4573

**Reply by bobbrow** [reply]
@sean-mcmanus, in this case it looks like CMake Tools doesn't find the configuration to send to cpptools.  There may still be other issues with cpptools and symlinks after that, but I think we need a bug open for CMake Tools to investigate as well.

@stark-dev could you open a bug in the CMake Tools repo with information about how your symlinks work and how you open the folder so we can debug it?  https://github.com/microsoft/vscode-cmake-tools/issues/new/choose

**Reply by stark-dev** [reply]
Sure, I will upload the logs and open a bug on CMake Tools, thanks a lot for your help!

**Reply by stark-dev** [reply]
https://github.com/microsoft/vscode-cmake-tools/issues/2713

**Comment by bobbrow** [comment, marked as answer]
In case anyone else hits this issue, the root cause is a discrepancy between how symlinks are handled.  Refer to [this discussion](https://github.com/microsoft/vscode-discussions/discussions/122).  cpptools sends CMake Tools the resolved paths to files (`realpath`), but CMake Tools is matching them against the unresolved paths and fails.

The workaround for now is to launch VS Code from inside the linked path using `code .`  Currently VS Code will report the workspace folder as the `realpath` in this case and CMake Tools won't get confused.

**Reply by stark-dev** [reply]
Thanks again for your help!
